### DISCUSSION TITLE
ERROR: Could not build wheels which use PEP 517 and cannot be installed directly

### DISCUSSION THREAD

**Discussion by JulianOrteil** [pyscaffold/pyscaffold#548] [Q&A] [answered]
I am building a library that uses pretty much default configurations from pyscaffold. Sometime this morning, I suddenly started getting the following error when I attempt to run my tests and I am stumped as to the root cause.

```
Building wheels for collected packages: python-step-series
  Building wheel for python-step-series (PEP 517): started
  Building wheel for python-step-series (PEP 517): finished with status 'done'
  Created wheel for python-step-series: filename=python_step_series-0.0.2.dev1.post1.dev0_g1f5fc90.d20211206-py3-none-any.whl size=17943 sha256=6db044ba1828101d1ecd7f27be55a42b63711d537a6ea01d4935eaa4122655ef
  Stored in directory: /home/user/.cache/pip/wheels/46/6f/be/601a51d80e3889cae97cbe8b1a98cc4be1004f1e97546d94ef
  WARNING: Built wheel for python-step-series is invalid: Metadata 1.2 mandates PEP 440 version, but '0.0.2.dev1.post1.dev0-g1f5fc90.d20211206' is not
Failed to build python-step-series
ERROR: Could not build wheels for python-step-series which use PEP 517 and cannot be installed directly
```

I know it has something to do with git tagging and the fact that my working tree is dirty, but how to fix this is beyond me. Until this morning, I never had an issue. Creating a new tag every time I want to run my tests (using tox) seems to band-aid this--until my working tree becomes dirty again. Regardless, this band-aid isn't efficient nor proper.

I'm using tox-conda instead of virtualenv, but each environment is reporting Pip version: 21.2.4 (python 3.7,3.8, 3.9. 3.10)

Has anyone else encountered or have knowledge of this issue and how to fix it?

Thanks,
Jules

**Comment by FlorianWilhelm** [comment, marked as answer]
Hi Jules, thanks for reporting this problem. I just tried to reproduce it:
```
conda create -n new_env python=3.10
activate new_env
conda install pyscaffold tox
putup test_proj
cd test_proj
# added a file, committed, changed it to make the repo dirty
python setup.py --version # returns `0.0.post1.dev2+ge87ef71.d20211206`
tox -e build # works
tox # works
```
Could you please try to reproduce it in a similar manner? Please also provide the output of `conda env export`. 

Cheers, Florian

**Reply by JulianOrteil** [reply]
Hi Florian, thanks for your quick reply. I must apologize for wasting your time. I decided to wipe the environment a third time while also deleting the cached pip wheels and start from scratch again. Apparently that worked as I no longer see this issue coming up. Everything is running fine again, but thank you for being so quick.

**Reply by JulianOrteil** [reply]
As a wise man once said:
> There's only two states of programming: it doesn't work and you don't know why, and it works but you don't know why.

Certainly true in this case.

**Reply by FlorianWilhelm** [reply]
:-) No problem. Good that it works.

**Comment by FlorianWilhelm** [comment]
After some more googling, I think the problem is not related to the fact that non-tagged commits or dirty repos are not pep 440 conform. This only emits a warning.

The real culprit here might be the fact that you are somehow using an old pip or setuptools version. Check out this [Stackoverflow question](https://stackoverflow.com/questions/64038673/could-not-build-wheels-for-which-use-pep-517-and-cannot-be-installed-directly) for more details.
